### PR TITLE
Make Windows.isValid reject paths ending in a dot or space

### PR DESCRIPTION
**Pull request by ag-eitilt**
This is issue 2 of 3 following up on potential cross-platform compatibility problems that were identified in haskell/cabal#7429 (alongside haskell/cabal#7479 and haskell/cabal#7480).  That PR fixed a bug in the path validation state machine, but in doing so I realized that some of the Windows idiosyncrasies weren't being checked; the resulting discussion revealed that most of them were supposed to be checked elsewhere, but I volunteered to verify that anyway, once I got time.  I've since done so, and am now recording my results:

`Windows.isValid` currently has no issues with the inputs `test.` and `test `--note the trailing dot/space.  The discussion in #48 implies that this may only be an issue as the last segment of a path, though I just tried a `mkdir test.\\test` in a VM and got `test\\test`.  No matter what degree of stripping occurs, however, I would argue that the version with the trailing character(s) shouldn't be considered valid; whether the system rejects it entirely or opens a different file than expected, any users of this library won't get the result they expect from a file passing `isValid`.  As a concrete example, `cabal check` uses/will use that test to help ensure anything that gets uploaded to Hackage will (mostly) work on any supported system, but trying to open one of the packaged data files ending on a dot--which successfully passes muster on FreeBSD--fails on Windows.

This patch makes `Windows.isValid` reject a path containing any segments ending in a dot or a period.  I don't mean this to be a statement on whether or not `normalize` should strip those, but that should probably factor in to the decision on whether or not this PR should be merged.

### CODE REVIEW

**Comment by Mikolaj** [comment]
I can imagine use cases where the offending characters should be removed, but I agree for cabal we should reject such paths instead, e.g., due to the data files paths you mention.

Is there any other decision to take?

**Comment by hasufell** [comment]
It appears that the windows documentation agrees with you: https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions

> Do not end a file or directory name with a space or a period. Although the underlying file system may support such names, the Windows shell and user interface does not. However, it is acceptable to specify a period as the first character of a name. For example, ".temp".

I believe we should improve the documentation of `isValid` to indicate what constraints we're really following and make clear that these are not exhaustive. It's a best-effort.

File: System/FilePath/Internal.hs [hunk 1]
```diff
@@ -920,6 +920,10 @@ badElements =
 -- > Windows: isValid "\\\\\\foo" == False
 -- > Windows: isValid "\\\\?\\D:file" == False
 -- > Windows: isValid "foo\tbar" == False
+-- > Windows: isValid "foo." == False
+-- > Windows: isValid "foo " == False
+-- > Windows: isValid "foo./bar" == False
+-- > Windows: isValid "foo /bar" == False
```
Comment (hasufell) [inline System/FilePath/Internal.hs:926]:
```suggestion
-- > Windows: isValid "foo.\\bar" == False
-- > Windows: isValid "foo \\bar" == False
```

Should probably use the primary path separator here.

File: tests/TestGen.hs [hunk 2]
```diff
@@ -422,6 +422,10 @@ tests =
     ,("W.isValid \"\\\\\\\\\\\\foo\" == False", property $ W.isValid "\\\\\\foo" == False)
     ,("W.isValid \"\\\\\\\\?\\\\D:file\" == False", property $ W.isValid "\\\\?\\D:file" == False)
     ,("W.isValid \"foo\\tbar\" == False", property $ W.isValid "foo\tbar" == False)
+    ,("W.isValid \"foo.\" == False", property $ W.isValid "foo." == False)
+    ,("W.isValid \"foo \" == False", property $ W.isValid "foo " == False)
+    ,("W.isValid \"foo./bar\" == False", property $ W.isValid "foo./bar" == False)
+    ,("W.isValid \"foo /bar\" == False", property $ W.isValid "foo /bar" == False)
```
Comment (hasufell) [inline tests/TestGen.hs:428]:
```suggestion
    ,("W.isValid \"foo.\\bar\" == False", property $ W.isValid "foo.\\bar" == False)
    ,("W.isValid \"foo \\bar\" == False", property $ W.isValid "foo \\bar" == False)
```

**Comment by hasufell** [comment]
Although we don't change any types... we're changing the behavior of a function. PVP isn't particularly clear whether this is a breaking change or not, but I believe that's in the scope of a maintainer?

It's also not entirely clear whether this is a "bug". Maybe someone started relying on it. Bumping A.B as per PVP here will maybe force users to read the changelog? Opinions?

**Comment by Mikolaj** [comment]
My gut feeling is to mark this as a breaking change.

**Comment by hasufell** [comment]
Merged to https://github.com/haskell/filepath/tree/1.5.0.0

**Pull request closed without merging**